### PR TITLE
Actionchain improvements

### PR DESCRIPTION
**Pull request by manasdk**
- Proper schema for ActionChain
- Schema validation catches malformed yaml/json
- default node is now optional
- support for vars
```
{
    "vars": {
        "strtype": "{{system.a}}",
        "inttype": 1
    },
    "chain": [
        {
            "name": "c1",
            "ref": "wolfpack.a2",
            "params":
            {
                "inttype": "{{inttype}}",
                "strtype": "{{strtype}}",
                "booltype": true
            }
        }
    ],
    "default": "c1"
}
```
- support for output variable per action_node
```
{
    "chain": [
        {
            "name": "c1",
            "ref": "wolfpack.a2",
            "params":
            {
                "inttype": "{{inttype}}",
                "strtype": "{{strtype}}",
                "booltype": true
            },
            "output":
            {
                "o1": "{{c1.foo.bar}}"
            },
            "on-success": "c2"
        },
        {
            "name": "c2",
            "ref": "wolfpack.a2",
            "params":
            {
                "strtype": "{{o1}}"
            }
        }
    ],
    "default": "c1"
}
```
Note : Full result of `c1` is still available and variables published by `c1` are in the global context.

### CODE REVIEW

File: st2common/st2common/models/system/actionchain.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                "required": True
+            },
+            "ref": {
+                "type": "string",
+                "description": "Ref of the action to be executed.",
+                "required": True
+            },
+            "params": {
+                "type": "object",
+                "description": "query context to be used by querier.",
```
Comment (Kami) [inline st2common/st2common/models/system/actionchain.py]: This should probably say params which are passed to the action, right?

Comment (manasdk) [inline st2common/st2common/models/system/actionchain.py]: Copy-paste error. Will fix.

**Comment by dzimine** [comment]
* Strongly suggest to  change "params" to "input".  We tried many variations and in the context of workflow, input/output is probably the best. 

Are we borrowing type names from somewhere or inventing them? `strtype, inttype, booltype`? I am concerned that we use different wording for types here vs action definition.

**Comment by manasdk** [comment]
> Are we borrowing type names from somewhere or inventing them? strtype, inttype, booltype? I am concerned that we use different wording for types here vs action definition.

That is only an example. I am using examples from tests and used those variable names to help debugging tests simpler and also to avoid having to come up with clever values. To be clear these are user defined and not type names of any sort. The types we support are as per jsonschema.

See http://json-schema.org/latest/json-schema-core.html#anchor8

**Comment by dzimine** [comment]
> That is only an example. I am using examples
ok. thanks.

**Comment by manasdk** [comment]
> "params" to "input"

I guess we should have kept you in the loop for the original schema as well! 

Anyway, nomenclature wise I am ok with either `params` or `input`. However, changing it now implies breaking backwards compatability. Even @lakshmi-kannan pointed out that ActionChain uses parameters and mistral uses input so in order to avoid confusion I see the value of using input.

Having said that I must point out few counter points -
1. In the st2 system we do not use `input` anywhere. It is some variation of parameters.
    https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/api/action.py#L78
    https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/api/action.py#L155
    https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/api/action.py#L233
    https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/api/rule.py#L103
    Before we switch to `input` we must consider the lack of internal consistency.
2. I am more comfortable with the paradigm that a function has parameters and is supplied input at runtime. But, the language tends to get a little murky. I guess at execution time input is the accurate terms but then again it is nitpicking. In anycase the action_node does not seem like a function call more more of a definition so params felt ok.

My counter point is only to highlight that the naming isn't arbitrary but if there is enough of discussion already over `input` and `params` then I would hate to have it all over again.

**Comment by jfryman** [comment]
> Before we switch to input we must consider the lack of internal consistency.

I was actually going to mention this. Within AC, `params` != `parameters`. Now, I realize words mean something, but just the fact that the same usage is used (even slightly differently) is yet another thing a new user needs to keep in their head. I would rather use the wrong word everywhere, than have multiple right words.

$0.02

**Comment by jfryman** [comment]
> Before we switch to input we must consider the lack of internal consistency.

I was actually going to mention this, but the disconnect between AC's `params` and `parameters` used everywhere else in the codebase is noticeably annoying when developing.  Now, I realize words mean something, but just the fact that the same usage is used (even slightly differently) is yet another thing a new user needs to keep in their head. I would rather use the wrong word everywhere, than have multiple right words.

If you are going to change `params` to anything, I would strongly suggest `parameters`. I do realize that `mistral` uses the `input`/`output` verbs, but that is an integration versus a native component. Because of this, I am inclined to say that the delta in verbs is okay... it's only natural that external tools will model things slightly differently. However, AC is native, and I think internal consistency is merited in this case. 

$0.02

**Comment by manasdk** [comment]
>> Note : Full result of c1 is no longer available since c1 is publishing output.

>Also curious about this behavior. What's your thought here?

One way I thought about this was that choosing to publish an output variable is an indication of which variable is required in the output. Also, from an implementation perspective avoids having to deal with overrides.

TBH I am conflicted about this choice myself and go pick either one of these options -
1. Only publish output
2. Publish output + full result

**Comment by DoriftoShoes** [comment]
I want to be consistent internally and I do not want to switch to 'input', I actually hate the word 'input' in Mistral.  I can guarantee that I will mess up the verbiage when talking to people everyday...and if I (someone who uses it everyday) can't keep it straight, other people won't either.

On the published results dealio.  As long as I have the option of NOT specifying what to publish and still getting access to the full results body throughout the workflow this is fine.  In most places using `set` in Jinja is enough for me.

**Comment by manasdk** [comment]
> the disconnect between AC's `params` and `parameters`

Yeah. I messed that one up. It should've been `parameters` across the board.

**Comment by jfryman** [comment]
> On the published results dealio. As long as I have the option of NOT specifying what to publish and still getting access to the full results body throughout the workflow this is fine. In most places using set in Jinja is enough for me.

That seems to be a fair in-between. If you specify output, that's it... otherwise you get the raw return. Would this paint us into any corners? This seems to be closer to @manasdk option 2...

**Comment by manasdk** [comment]
>  If you specify output, that's it... otherwise you get the raw return. 

That is the current impl. revised options since I left out details -

1. publish output **or** publish full result. [1 or the other]
2. publish output  **and** publish full result. [1 or the other]

Current impl is option 1.

**Comment by jfryman** [comment]
@manasdk Got it. :+1: I'd leave as-is then. Option 2 could lead to some nasty code... at least Option 1 sort of encourages good behavior. I appreciate the slight detour with explanation. :heart:

**Comment by manasdk** [comment]
> I appreciate the slight detour with explanation.

NP. I shall revise the description on the schema as this was obviously unclear at first read.

**Comment by dzimine** [comment]
Discussed with Manas, decided
1) publish publishes to global, not to local
2) when task publishes, the task action result is STILL accessible via `{{taskname.output.reference}}`
3) output -> publish

Here is the YAML edition: 
```
vars:
  - strvar: {{system.a}}
  - intvar: 1
chain:
  - 
  name: c1
  ref: wolfpack.a2
  params: 
    intpar: {{intvar}} # refer global scope variable
    strpar: {{strvar}}
    boolvar: true
  publish:
    pubvar1: "{{c1.foo.bar}}" # refer the action output by task name 
  on-success: c2
  -
  name: c2
  ref: wolfpack.a2 
  params:
    intvar: {{c1.foo.bar}} #refer 
    strvar: {{pubvar1}} # previously published is in global scope
  publish:  
    pubvar2:                # mix and match
       k1: "{{c1.foo.bar}}  # results of this task
       k2: {{intvar}}       # with global vars
       k3: {{c2.bar.baz}}   # and direct references to prev tasks' results

# default: c1 # explicit default is optional
```

**Comment by manasdk** [comment]
Couple of updates -

(1) `output` renamed to `publish`
(2) published variables are now only available in context of actionchain. They are not published to the result of the chain.
(3) complete result of each execution is still available.
(4) for now only strings can be published. We can soon extend this to include publishing of other structured data.

**Pull request merged**